Basic experimental fog code extracted from jshexen2 (activate with
"gl_fog 1", see the cvars for other options):
- underwater fogging removed
- no menu entries in Options :: OpenGL Features, yet
- slightly cleaned-up according to original quake code
- there is some striped view in fogged areas ;(
- seems NOT working correctly without multitexturing
- obviously still needs fixing ;)  the only reason I
  revied this crap is that the glowed missiles look
  somewhat better in fog (personal liking), sigh...

diff -urNp uhexen2-20060311/hexen2/gl_dl_rlight.c uhexen2-20060312/hexen2/gl_dl_rlight.c
--- uhexen2-20060311/hexen2/gl_dl_rlight.c	2006-03-12 00:53:29.000000000 -0500
+++ uhexen2-20060312/hexen2/gl_dl_rlight.c	2006-03-12 01:13:39.000000000 -0500
@@ -117,6 +117,10 @@ void R_RenderDlights (void)
 
 	r_dlightframecount = r_framecount + 1;	// because the count hasn't
 						//  advanced yet for this frame
+
+	// disable drawing fog on lights
+	if (gl_fogenable.value)
+		glDisable_fp(GL_FOG);
 	glDepthMask_fp (0);
 	glDisable_fp (GL_TEXTURE_2D);
 	glShadeModel_fp (GL_SMOOTH);
@@ -136,6 +140,9 @@ void R_RenderDlights (void)
 	glEnable_fp (GL_TEXTURE_2D);
 	glBlendFunc_fp (GL_SRC_ALPHA, GL_ONE_MINUS_SRC_ALPHA);
 	glDepthMask_fp (1);
+	// re-enable fog
+	if (gl_fogenable.value)
+		glEnable_fp(GL_FOG);
 }
 
 
diff -urNp uhexen2-20060311/hexen2/gl_dl_rmain.c uhexen2-20060312/hexen2/gl_dl_rmain.c
--- uhexen2-20060311/hexen2/gl_dl_rmain.c	2006-03-11 17:51:16.000000000 -0500
+++ uhexen2-20060312/hexen2/gl_dl_rmain.c	2006-03-12 01:24:36.000000000 -0500
@@ -33,6 +33,8 @@ mplane_t	*mirror_plane;
 
 static float	model_constant_alpha;
 
+static GLfloat	fog_colours[4];
+
 static float	r_time1;
 static float	r_lasttime1 = 0;
 
@@ -96,6 +98,14 @@ cvar_t	gl_stencilshadow = {"gl_stencilsh
 cvar_t	gl_glows = {"gl_glows","1",true};
 cvar_t	gl_other_glows = {"gl_other_glows","0",true};
 cvar_t	gl_missile_glows = {"gl_missile_glows","1",true};
+cvar_t	gl_fogenable = {"gl_fog", "0", true};
+cvar_t	gl_fogstart = {"gl_fogstart", "400.0", true};
+cvar_t	gl_fogend = {"gl_fogend", "2000.0", true};
+cvar_t	gl_fogdensity = {"gl_fogdensity", "0.2", true};
+cvar_t	gl_fogred = {"gl_fogred","0.5", true};
+cvar_t	gl_foggreen = {"gl_foggreen","0.5", true};
+cvar_t	gl_fogblue = {"gl_fogblue","0.5", true};
+cvar_t	gl_fogalpha = {"gl_fogalpha", "0.5", true};
 
 cvar_t	gl_coloredlight = {"gl_coloredlight","0",true};
 cvar_t	gl_colored_dynamic_lights = {"gl_colored_dynamic_lights","0",true};
@@ -1297,6 +1307,9 @@ static void R_DrawAllGlows (void)
 	if (!r_drawentities.value)
 		return;
 
+	if (gl_fogenable.value)
+		glDisable_fp(GL_FOG);
+
 	glDepthMask_fp (0);
 	glDisable_fp (GL_TEXTURE_2D);
 	glShadeModel_fp (GL_SMOOTH);
@@ -1317,6 +1330,9 @@ static void R_DrawAllGlows (void)
 		}
 	}
 
+	if (gl_fogenable.value)
+		glEnable_fp(GL_FOG);
+
 	glDisable_fp (GL_BLEND);
 	glEnable_fp (GL_TEXTURE_2D);
 	glBlendFunc_fp (GL_SRC_ALPHA, GL_ONE_MINUS_SRC_ALPHA);
@@ -1736,6 +1752,23 @@ static void R_RenderScene (void)
 
 	R_MarkLeaves ();	// done here so we know if we're in water
 
+#if 0
+// WHY DID JSHEXEN2 PUT THIS HERE ??!!
+	if (gl_fogenable.value)
+	{
+		glFogi_fp(GL_FOG_MODE, GL_LINEAR);
+
+		fog_colours[0] = gl_fogred.value;
+		fog_colours[1] = gl_foggreen.value;
+		fog_colours[2] = gl_fogblue.value;
+
+		glFogfv_fp(GL_FOG_COLOR, fog_colours);
+		glFogf_fp(GL_FOG_START, gl_fogstart.value);
+		glFogf_fp(GL_FOG_END, gl_fogend.value);
+		glEnable_fp(GL_FOG);
+	}
+#endif
+
 	R_DrawWorld ();		// adds static entities to the list
 
 	S_ExtraUpdate ();	// don't let sound get messed up if going slow
@@ -1744,6 +1777,8 @@ static void R_RenderScene (void)
 
 	R_DrawAllGlows();
 
+	//glDisable_fp(GL_FOG);	// already done in R_RenderDlights()
+
 	R_RenderDlights ();
 
 #ifdef GLTEST
@@ -1941,6 +1976,39 @@ void R_RenderView (void)
 
 	R_Clear ();
 
+	if (gl_fogenable.value)
+	{
+		if (gl_fogstart.value < 5)
+			Cvar_SetValue ("gl_fogstart", 5);
+		if (gl_fogstart.value > 4095)
+			Cvar_SetValue ("gl_fogstart", 4095);
+		if (gl_fogend.value < 6)
+			Cvar_SetValue ("gl_fogend", 6);
+		if (gl_fogend.value > 4096)
+			Cvar_SetValue ("gl_fogend", 4096);
+		if (gl_fogstart.value >= gl_fogend.value)
+			Cvar_SetValue ("gl_fogend", gl_fogstart.value + 1);
+
+		fog_colours[0] = (GLfloat) gl_fogred.value;
+		fog_colours[1] = (GLfloat) gl_foggreen.value;
+		fog_colours[2] = (GLfloat) gl_fogblue.value;
+		fog_colours[3] = (GLfloat) 1;//gl_fogalpha.value;
+
+		glFogi_fp (GL_FOG_MODE, GL_LINEAR);
+	//	glFogi_fp (GL_FOG_MODE, GL_EXP);
+	//	glFogi_fp (GL_FOG_MODE, GL_EXP2);
+		glFogfv_fp (GL_FOG_COLOR, fog_colours);
+		glFogf_fp (GL_FOG_START, gl_fogstart.value);
+		glFogf_fp (GL_FOG_END, gl_fogend.value);
+		glFogf_fp (GL_FOG_DENSITY, gl_fogdensity.value);
+	//	glHint_fp (GL_FOG_HINT, GL_NICEST);
+		glEnable_fp (GL_FOG);
+	}
+	else
+	{
+		glDisable_fp (GL_FOG);
+	}
+
 	// render normal view
 	R_RenderScene ();
 
@@ -1956,6 +2024,9 @@ void R_RenderView (void)
 
 	R_DrawViewModel();
 
+	if (gl_fogenable.value)
+		glDisable_fp(GL_FOG);
+
 	// render mirror view
 	R_Mirror ();
 
@@ -1963,6 +2034,8 @@ void R_RenderView (void)
 
 	if (r_speeds.value)
 		R_PrintTimes ();
+
+	glDepthMask_fp(1);
 }
 
 
diff -urNp uhexen2-20060311/hexen2/gl_dl_rmisc.c uhexen2-20060312/hexen2/gl_dl_rmisc.c
--- uhexen2-20060311/hexen2/gl_dl_rmisc.c	2006-03-12 00:53:29.000000000 -0500
+++ uhexen2-20060312/hexen2/gl_dl_rmisc.c	2006-03-12 01:30:24.000000000 -0500
@@ -247,6 +247,15 @@ void R_Init (void)
 	Cvar_RegisterVariable (&gl_other_glows);
 	Cvar_RegisterVariable (&gl_stencilshadow);
 
+	Cvar_RegisterVariable (&gl_fogenable);
+	Cvar_RegisterVariable (&gl_fogstart);
+	Cvar_RegisterVariable (&gl_fogend);
+	Cvar_RegisterVariable (&gl_fogdensity);
+	Cvar_RegisterVariable (&gl_fogalpha);
+	Cvar_RegisterVariable (&gl_fogred);
+	Cvar_RegisterVariable (&gl_fogblue);
+	Cvar_RegisterVariable (&gl_foggreen);
+
 	Cvar_RegisterVariable (&gl_coloredlight);
 	Cvar_RegisterVariable (&gl_colored_dynamic_lights);
 	Cvar_RegisterVariable (&gl_extra_dynamic_lights);
diff -urNp uhexen2-20060311/hexen2/gl_func.h uhexen2-20060312/hexen2/gl_func.h
--- uhexen2-20060311/hexen2/gl_func.h	2006-01-16 11:43:22.000000000 -0500
+++ uhexen2-20060312/hexen2/gl_func.h	2006-03-12 01:13:39.000000000 -0500
@@ -47,6 +47,11 @@ GL_FUNCTION(void, glColor3ubv, (const GL
 GL_FUNCTION(void, glColor3f, (GLfloat,GLfloat,GLfloat))
 GL_FUNCTION(void, glClearColor, (GLclampf,GLclampf,GLclampf,GLclampf))
 
+GL_FUNCTION(void, glFogf, (GLenum,GLfloat))
+GL_FUNCTION(void, glFogfv, (GLenum, const GLfloat *))
+GL_FUNCTION(void, glFogi, (GLenum,GLint))
+GL_FUNCTION(void, glFogiv, (GLenum, const GLint *))
+
 GL_FUNCTION(void, glAlphaFunc, (GLenum,GLclampf))
 GL_FUNCTION(void, glBlendFunc, (GLenum,GLenum))
 GL_FUNCTION(void, glShadeModel, (GLenum))
diff -urNp uhexen2-20060311/hexen2/glquake.h uhexen2-20060312/hexen2/glquake.h
--- uhexen2-20060311/hexen2/glquake.h	2006-03-11 17:51:16.000000000 -0500
+++ uhexen2-20060312/hexen2/glquake.h	2006-03-12 01:27:06.000000000 -0500
@@ -360,6 +360,16 @@ extern	cvar_t	gl_glows;
 extern	cvar_t	gl_other_glows;
 extern	cvar_t	gl_missile_glows;
 
+extern	cvar_t	gl_fogenable;
+extern	cvar_t	gl_fogstart;
+extern	cvar_t	gl_fogend;
+extern	cvar_t	gl_fogdensity;
+extern	cvar_t	gl_fogalpha;
+extern	cvar_t	gl_fogred;
+extern	cvar_t	gl_fogblue;
+extern	cvar_t	gl_foggreen;
+extern	cvar_t	gl_underwaterfog;
+
 extern	cvar_t	gl_coloredlight;
 extern	cvar_t	gl_colored_dynamic_lights;
 extern	cvar_t	gl_extra_dynamic_lights;
diff -urNp uhexen2-20060311/hexenworld/Client/gl_func.h uhexen2-20060312/hexenworld/Client/gl_func.h
--- uhexen2-20060311/hexenworld/Client/gl_func.h	2006-01-16 11:43:23.000000000 -0500
+++ uhexen2-20060312/hexenworld/Client/gl_func.h	2006-03-12 01:13:39.000000000 -0500
@@ -47,6 +47,11 @@ GL_FUNCTION(void, glColor3ubv, (const GL
 GL_FUNCTION(void, glColor3f, (GLfloat,GLfloat,GLfloat))
 GL_FUNCTION(void, glClearColor, (GLclampf,GLclampf,GLclampf,GLclampf))
 
+GL_FUNCTION(void, glFogf, (GLenum,GLfloat))
+GL_FUNCTION(void, glFogfv, (GLenum, const GLfloat *))
+GL_FUNCTION(void, glFogi, (GLenum,GLint))
+GL_FUNCTION(void, glFogiv, (GLenum, const GLint *))
+
 GL_FUNCTION(void, glAlphaFunc, (GLenum,GLclampf))
 GL_FUNCTION(void, glBlendFunc, (GLenum,GLenum))
 GL_FUNCTION(void, glShadeModel, (GLenum))
diff -urNp uhexen2-20060311/hexenworld/Client/glquake.h uhexen2-20060312/hexenworld/Client/glquake.h
--- uhexen2-20060311/hexenworld/Client/glquake.h	2006-03-11 17:51:20.000000000 -0500
+++ uhexen2-20060312/hexenworld/Client/glquake.h	2006-03-12 01:26:43.000000000 -0500
@@ -365,6 +365,16 @@ extern	cvar_t	gl_glows;
 extern	cvar_t	gl_other_glows;
 extern	cvar_t	gl_missile_glows;
 
+extern	cvar_t	gl_fogenable;
+extern	cvar_t	gl_fogstart;
+extern	cvar_t	gl_fogend;
+extern	cvar_t	gl_fogdensity;
+extern	cvar_t	gl_fogalpha;
+extern	cvar_t	gl_fogred;
+extern	cvar_t	gl_fogblue;
+extern	cvar_t	gl_foggreen;
+extern	cvar_t	gl_underwaterfog;
+
 extern	cvar_t	gl_coloredlight;
 extern	cvar_t	gl_colored_dynamic_lights;
 extern	cvar_t	gl_extra_dynamic_lights;
diff -urNp uhexen2-20060311/hexenworld/Client/gl_rlight.c uhexen2-20060312/hexenworld/Client/gl_rlight.c
--- uhexen2-20060311/hexenworld/Client/gl_rlight.c	2006-03-12 00:53:30.000000000 -0500
+++ uhexen2-20060312/hexenworld/Client/gl_rlight.c	2006-03-12 01:13:39.000000000 -0500
@@ -139,6 +139,10 @@ void R_RenderDlights (void)
 
 	r_dlightframecount = r_framecount + 1;	// because the count hasn't
 						//  advanced yet for this frame
+
+	// disable drawing fog on lights
+	if (gl_fogenable.value)
+		glDisable_fp(GL_FOG);
 	glDepthMask_fp (0);
 	glDisable_fp (GL_TEXTURE_2D);
 	glShadeModel_fp (GL_SMOOTH);
@@ -158,6 +162,9 @@ void R_RenderDlights (void)
 	glEnable_fp (GL_TEXTURE_2D);
 	glBlendFunc_fp (GL_SRC_ALPHA, GL_ONE_MINUS_SRC_ALPHA);
 	glDepthMask_fp (1);
+	// re-enable fog
+	if (gl_fogenable.value)
+		glEnable_fp(GL_FOG);
 }
 
 
diff -urNp uhexen2-20060311/hexenworld/Client/gl_rmain.c uhexen2-20060312/hexenworld/Client/gl_rmain.c
--- uhexen2-20060311/hexenworld/Client/gl_rmain.c	2006-03-11 17:51:20.000000000 -0500
+++ uhexen2-20060312/hexenworld/Client/gl_rmain.c	2006-03-12 01:25:42.000000000 -0500
@@ -33,6 +33,8 @@ mplane_t	*mirror_plane;
 
 static float	model_constant_alpha;
 
+static GLfloat	fog_colours[4];
+
 static float	r_time1;
 static float	r_lasttime1 = 0;
 
@@ -101,6 +103,14 @@ cvar_t	gl_stencilshadow = {"gl_stencilsh
 cvar_t	gl_glows = {"gl_glows","1",true};
 cvar_t	gl_other_glows = {"gl_other_glows","0",true};
 cvar_t	gl_missile_glows = {"gl_missile_glows","1",true};
+cvar_t	gl_fogenable = {"gl_fog", "0", true};
+cvar_t	gl_fogstart = {"gl_fogstart", "400.0", true};
+cvar_t	gl_fogend = {"gl_fogend", "2000.0", true};
+cvar_t	gl_fogdensity = {"gl_fogdensity", "0.2", true};
+cvar_t	gl_fogred = {"gl_fogred","0.5", true};
+cvar_t	gl_foggreen = {"gl_foggreen","0.5", true};
+cvar_t	gl_fogblue = {"gl_fogblue","0.5", true};
+cvar_t	gl_fogalpha = {"gl_fogalpha", "0.5", true};
 
 cvar_t	gl_coloredlight = {"gl_coloredlight","0",true};
 cvar_t	gl_colored_dynamic_lights = {"gl_colored_dynamic_lights","0",true};
@@ -1299,6 +1309,9 @@ static void R_DrawAllGlows (void)
 	if (!r_drawentities.value)
 		return;
 
+	if (gl_fogenable.value)
+		glDisable_fp(GL_FOG);
+
 	glDepthMask_fp (0);
 	glDisable_fp (GL_TEXTURE_2D);
 	glShadeModel_fp (GL_SMOOTH);
@@ -1319,6 +1332,9 @@ static void R_DrawAllGlows (void)
 		}
 	}
 
+	if (gl_fogenable.value)
+		glEnable_fp(GL_FOG);
+
 	glDisable_fp (GL_BLEND);
 	glEnable_fp (GL_TEXTURE_2D);
 	glBlendFunc_fp (GL_SRC_ALPHA, GL_ONE_MINUS_SRC_ALPHA);
@@ -1796,6 +1812,23 @@ static void R_RenderScene (void)
 
 	R_MarkLeaves ();	// done here so we know if we're in water
 
+#if 0
+// WHY DID JSHEXEN2 PUT THIS HERE ??!!
+	if (gl_fogenable.value)
+	{
+		glFogi_fp(GL_FOG_MODE, GL_LINEAR);
+
+		fog_colours[0] = gl_fogred.value;
+		fog_colours[1] = gl_foggreen.value;
+		fog_colours[2] = gl_fogblue.value;
+
+		glFogfv_fp(GL_FOG_COLOR, fog_colours);
+		glFogf_fp(GL_FOG_START, gl_fogstart.value);
+		glFogf_fp(GL_FOG_END, gl_fogend.value);
+		glEnable_fp(GL_FOG);
+	}
+#endif
+
 	R_DrawWorld ();		// adds static entities to the list
 
 	S_ExtraUpdate ();	// don't let sound get messed up if going slow
@@ -1804,6 +1837,8 @@ static void R_RenderScene (void)
 
 	R_DrawAllGlows();
 
+	//glDisable_fp(GL_FOG);	// already done in R_RenderDlights()
+
 	R_RenderDlights ();
 
 #ifdef GLTEST
@@ -2060,6 +2095,39 @@ void R_RenderView (void)
 
 	R_Clear ();
 
+	if (gl_fogenable.value)
+	{
+		if (gl_fogstart.value < 5)
+			Cvar_SetValue ("gl_fogstart", 5);
+		if (gl_fogstart.value > 4095)
+			Cvar_SetValue ("gl_fogstart", 4095);
+		if (gl_fogend.value < 6)
+			Cvar_SetValue ("gl_fogend", 6);
+		if (gl_fogend.value > 4096)
+			Cvar_SetValue ("gl_fogend", 4096);
+		if (gl_fogstart.value >= gl_fogend.value)
+			Cvar_SetValue ("gl_fogend", gl_fogstart.value + 1);
+
+		fog_colours[0] = (GLfloat) gl_fogred.value;
+		fog_colours[1] = (GLfloat) gl_foggreen.value;
+		fog_colours[2] = (GLfloat) gl_fogblue.value;
+		fog_colours[3] = (GLfloat) 1;//gl_fogalpha.value;
+
+		glFogi_fp (GL_FOG_MODE, GL_LINEAR);
+	//	glFogi_fp (GL_FOG_MODE, GL_EXP);
+	//	glFogi_fp (GL_FOG_MODE, GL_EXP2);
+		glFogfv_fp (GL_FOG_COLOR, fog_colours);
+		glFogf_fp (GL_FOG_START, gl_fogstart.value);
+		glFogf_fp (GL_FOG_END, gl_fogend.value);
+		glFogf_fp (GL_FOG_DENSITY, gl_fogdensity.value);
+	//	glHint_fp (GL_FOG_HINT, GL_NICEST);
+		glEnable_fp (GL_FOG);
+	}
+	else
+	{
+		glDisable_fp (GL_FOG);
+	}
+
 	// render normal view
 	R_RenderScene ();
 
@@ -2075,6 +2143,9 @@ void R_RenderView (void)
 
 	R_DrawViewModel();
 
+	if (gl_fogenable.value)
+		glDisable_fp(GL_FOG);
+
 	// render mirror view
 //	R_Mirror ();
 
@@ -2082,5 +2153,7 @@ void R_RenderView (void)
 
 	if (r_speeds.value)
 		R_PrintTimes ();
+
+	glDepthMask_fp(1);
 }
 
diff -urNp uhexen2-20060311/hexenworld/Client/gl_rmisc.c uhexen2-20060312/hexenworld/Client/gl_rmisc.c
--- uhexen2-20060311/hexenworld/Client/gl_rmisc.c	2006-03-12 00:53:30.000000000 -0500
+++ uhexen2-20060312/hexenworld/Client/gl_rmisc.c	2006-03-12 01:30:49.000000000 -0500
@@ -232,6 +232,15 @@ void R_Init (void)
 	Cvar_RegisterVariable (&gl_other_glows);
 	Cvar_RegisterVariable (&gl_stencilshadow);
 
+	Cvar_RegisterVariable (&gl_fogenable);
+	Cvar_RegisterVariable (&gl_fogstart);
+	Cvar_RegisterVariable (&gl_fogend);
+	Cvar_RegisterVariable (&gl_fogdensity);
+	Cvar_RegisterVariable (&gl_fogalpha);
+	Cvar_RegisterVariable (&gl_fogred);
+	Cvar_RegisterVariable (&gl_fogblue);
+	Cvar_RegisterVariable (&gl_foggreen);
+
 	Cvar_RegisterVariable (&gl_coloredlight);
 	Cvar_RegisterVariable (&gl_colored_dynamic_lights);
 	Cvar_RegisterVariable (&gl_extra_dynamic_lights);

